(PDB-1531) Add mbeans handler and ability to configure multiple registries

ajroetker opened the pull request:
This PR adds the ability to use multiple registries with tk-metrics by adding a protocol function that "upserts" registries on calls to `(get-metrics-registry <name of registry> <opts>)`

ajroetker commented:
@cprice404 I had some time today and was playing around with how we might add support for multiple registries to tk-metrics. Does this approach look reasonably sane? If not, what were you thinking? If it looks alright I'll touch up the PR :)

cprice404 commented on src/puppetlabs/trapperkeeper/services/metrics/metrics_service.clj:
```diff
@@ -4,14 +4,30 @@
             [puppetlabs.trapperkeeper.services.metrics.metrics-core :as core]
             [puppetlabs.trapperkeeper.services :as tk-services]))
 
+(def registries (atom {}))
```
global shouldn't be necessary

cprice404 commented:
@ajroetker this looks like it's heading in the right direction to me, and I'd be happy to take a PR for this if you have time to flesh it out!  Also would love to get the other metrics HTTP stuff that you mentioned.

ajroetker commented:
@cprice404 added the endpoint and fixed up the rest of the code

ajroetker commented:
Added some tests for the endpoint

cprice404 commented on project.clj:
```diff
@@ -7,22 +7,38 @@
+  :dependencies [[org.clojure/clojure "1.7.0"]
+                 [clj-time "0.10.0"]
+                 [commons-codec "1.9"]
+                 [org.clojure/tools.macro "0.1.5"]
```
I presume some of these are to resolve transitive dependency version conflicts w/lein pedantic?  If so can we try to go ahead and add the leading/trailing comments as per https://confluence.puppetlabs.com/pages/viewpage.action?pageId=48956980

cprice404 commented on src/puppetlabs/trapperkeeper/services/protocols/metrics.clj:
```diff
@@ -2,5 +2,9 @@
+    [this registry opts]
+    "Provides access to a MetricsRegistry configured with ops where `registry`
+     is the key used to look up the registry. Specifing no `registry` will
+     return the default MetricsRegistry"))
```
What is the type for `registry`?  Docs on what `opts` can include?  A reference to a prismatic schema would be ideal.

cprice404 commented on src/puppetlabs/trapperkeeper/services/metrics/metrics_service.clj:
```diff
@@ -6,12 +6,33 @@
+        :registry))
+
+  (get-metrics-registry [this registry opts]
+    (let [{:keys [registries]} (tk-services/service-context this)]
```
It'd be sweet if you could push some or all of the rest of this function into the core namespace and use schema on the fns to make it easier to intuit what types are legal for `registry` and `opts`.

cprice404 commented on test/puppetlabs/trapperkeeper/services/metrics/metrics_service_test.clj:
```diff
@@ -1,16 +1,42 @@
+          (is (= 200 (:status resp)))
+          (doseq [[metric path] body
+                  :let [resp (http-client/get (str "http://localhost:8180/metrics/v1" path))]]
+            (is (= 200 (:status resp)))))))))
```
Any basic assertions we could make against the body payloads?

cprice404 commented:
@ajroetker I think I'm basically good on this, except for minor nits above, plus docs.

One other question / thought that I have, though, is that it might be worth considering having the web service be a separate service from the existing metrics service.  That way, people could opt-in, and they wouldn't necessarily have to bring in the Jetty dependency if they didn't want it.

That said, I'm not sure I can think of a meaningful use case where people *wouldn't* want it, and I don't think there's any places where anyone at PL is using it today where this approach would cause any kind of problem.  But maybe it's worth circulating an e-mail to the dev list?

We'll also need to decide what the appropriate semver version for the next release of this will be if we ship it this way, with everything in the one service.

I think my preference is to keep it the way you have it, but just seems like a thing we might want to socialize a bit before making a final call.

ajroetker commented:
Still need to send out a message about having everything in one service @cprice404 but I think I addressed most of your other comments in here and added some more tests to make sure the other service function works as expected.

ajroetker commented:
@cprice404 separated the service for the endpoint and the core service

cprice404 commented:
Thanks, @ajroetker .  This LGTM.  My last request is for you to update the `index.md` file in the docs directory to reflect what you've done.  Maybe an example of using a domain, an example of enabling and/or using the web service, and removing the handful of places where I say "in the future we will have a web service!".  After that I'm +1 and thanks a ton for doing this.

Will see if I can rustle up a second reviewer now.

cprice404 commented:
@rlinehan @scottyw I know that you both already have a lot of things in flight, but since both of you have some recent experience with metrics-related stuff, if either of you had time in the next week or so to be a second reviewer for this I'd appreciate it!

scottyw commented on src/puppetlabs/trapperkeeper/services/metrics/ringutils.clj:
```diff
@@ -0,0 +1,10 @@
+  (:require [cheshire.core :as json]
+            [ring.util.response :as response]))
+
+(defn json-response [status body]
```
Could we use https://github.com/ring-clojure/ring-json?

scottyw commented on src/puppetlabs/trapperkeeper/services/metrics/metrics_core.clj:
```diff
@@ -18,36 +24,77 @@
-        (assoc :jmx-reporter (jmx-reporter registry jmx-config))))))
+(schema/defn get-or-initialize! :- RegistryContext
+  [config :- MetricsConfig
+   {:keys [registries] :as context} :- MetricsServiceContext
```
I think ```context``` isn't used

scottyw commented on src/puppetlabs/trapperkeeper/services/metrics/metrics_service.clj:
```diff
@@ -7,11 +7,40 @@
   (stop [this context]
-    (core/stop (tk-services/service-context this)))
+    (let [{:keys [registries] :as ctx} (tk-services/service-context this)]
+      (doseq [[_ metrics-reg] @registries]
```
Is there a race condition here if I call ```core/get-or-initialize!``` at the same time? Or does TK prevent this?

cprice404 replied:
If someone were to forego the service API and call into the core namespace directly, that could cause a problem, but I don't think that'd be considered a supported use case.

For downstream services using the supported API of calling into the service protocol functions to get access to the registry, we know they will be `stop`ped before this service's `stop` is called, based on how TK manages service dependencies / life cycles.  So I think this is pretty safe?

scottyw replied:
Agree that a direct call isn't a worry. I'm happy!

scottyw commented on src/puppetlabs/trapperkeeper/services/metrics/metrics_core.clj:
```diff
@@ -18,36 +24,77 @@
+   (comidi/wrap-routes
+    (comidi/context path
+        (comidi/context "/v1"
+            (comidi/context "/mbeans"
```
Could this just be the ```GET```?

ajroetker replied:
Think that might be a code alignment problem :) both the `GET`s fall under this context

scottyw commented:
Like it! Only real concern is the possible race. Thanks.

scottyw commented:
No race, I'm +1 to merge.

cprice404 commented:
cool, will merge once docs changes are in then.

ajroetker commented:
@cprice404 Pushed up the docs changes, let me know what you think!

cprice404 commented:
Thanks @ajroetker !  If you'd like to do a release and send out an announcement that'd be awesome, please feel free to ping in our chat room if you need any pointers on how to do that.  Most important thing is just to make sure the changelog is updated and the version number is correct.

cprice404 commented:
ping @ajroetker are you interested in / do you have bandwidth to do a release to get this into people's hands?
